File: frontend/lib/sales/client-segments.ts
/**
 * Client segment tags + duplicate match helpers (`2.3.2` / `SL-CLIENT-SEGMENTS-v1`).
 */

export const MAX_CLIENT_SEGMENTS = 32;
export const MAX_SEGMENT_NAME_LENGTH = 64;

export function normalizeSegmentName(raw: string): string {
  return raw.trim().replace(/\s+/g, " ");
}

export function mergeSegmentTags(current: string[], incoming: string): string[] {
  const name = normalizeSegmentName(incoming);
  if (!name) return current;
  if (name.length > MAX_SEGMENT_NAME_LENGTH) return current;
  const key = name.toLocaleLowerCase("ru");
  if (current.some((item) => item.toLocaleLowerCase("ru") === key)) {
    return current;
  }
  if (current.length >= MAX_CLIENT_SEGMENTS) return current;
  return [...current, name];
}

export function validateSegmentName(raw: string): string | null {
  const name = normalizeSegmentName(raw);
  if (!name) return "Укажите название сегмента";
  if (name.length > MAX_SEGMENT_NAME_LENGTH) {
    return "Сегмент не длиннее 64 символов";
  }
  return null;
}

export type ClientDuplicateCandidate = {
  id: number;
  company_name: string | null;
  contact_name: string;
  phone: string | null;
  inn: string | null;
  matched_on: string[];
};

export function duplicateMatchLabel(code: string): string {
  if (code === "name") return "имя";
  if (code === "phone") return "телефон";
  if (code === "inn") return "ИНН";
  return code;
}

File: backend/app/schemas/client_segments.py
from pydantic import BaseModel, Field, field_validator


def _blank_to_none(value: object) -> object:
    if isinstance(value, str):
        stripped = value.strip()
        return stripped or None
    return value


class ClientSegmentsReplace(BaseModel):
    tags: list[str] = Field(default_factory=list)

    @field_validator("tags")
    @classmethod
    def normalize_tags(cls, value: list[str]) -> list[str]:
        seen: set[str] = set()
        out: list[str] = []
        for raw in value:
            name = raw.strip()
            if not name:
                continue
            if len(name) > 64:
                raise ValueError("Segment name is at most 64 characters")
            key = name.casefold()
            if key in seen:
                continue
            seen.add(key)
            out.append(name)
        if len(out) > 32:
            raise ValueError("At most 32 segments per client")
        return out


class ClientDuplicateCandidate(BaseModel):
    id: int
    company_name: str | None = None
    contact_name: str
    phone: str | None = None
    inn: str | None = None
    matched_on: list[str] = Field(default_factory=list)
